vpr/src/place/place.cpp

```diff
@@ -2636,177 +3203,525 @@ static void update_bb(ClusterNetId net_id, t_bb* bb_coord_new, t_bb* bb_edge_new
+            update_bb_edge(net_id,
+                           bb_edge_new,
+                           bb_coord_new,
+                           bb_pin_sink_count_new,
+                           curr_bb_edge[layer_num].xmin,
+                           curr_bb_coord[layer_num].xmin,
+                           bb_edge_new[layer_num].xmin,
+                           bb_coord_new[layer_num].xmin);
+            if (bb_updated_before[net_id] == GOT_FROM_SCRATCH) {
+                return;
+            }
+        }
+
+        if (x_new > curr_bb_coord[layer_num].xmax) {
+            bb_edge_new[layer_num].xmax = 1;
+            bb_coord_new[layer_num].xmax = x_new;
+        } else if (x_new == curr_bb_coord[layer_num].xmax) {
+            bb_edge_new[layer_num].xmax = curr_bb_edge[layer_num].xmax + 1;
+            bb_coord_new[layer_num].xmax = curr_bb_coord[layer_num].xmax;
+        }
+    }
+
+    if (y_new < y_old) {
+        if (y_old == curr_bb_coord[layer_num].ymax) {
+            update_bb_edge(net_id,
+                           bb_edge_new,
+                           bb_coord_new,
+                           bb_pin_sink_count_new,
+                           curr_bb_edge[layer_num].ymax,
+                           curr_bb_coord[layer_num].ymax,
+                           bb_edge_new[layer_num].ymax,
+                           bb_coord_new[layer_num].ymax);
+            if (bb_updated_before[net_id] == GOT_FROM_SCRATCH) {
+                return;
+            }
+        }
+
+        if (y_new < curr_bb_coord[layer_num].ymin) {
+            bb_edge_new[layer_num].ymin = 1;
+            bb_coord_new[layer_num].ymin = y_new;
+        } else if (y_new == curr_bb_coord[layer_num].ymin) {
+            bb_edge_new[layer_num].ymin = curr_bb_edge[layer_num].ymin + 1;
+            bb_coord_new[layer_num].ymin = curr_bb_coord[layer_num].ymin;
+        }
+
+    } else if (y_new > y_old) {
+        if (y_old == curr_bb_coord[layer_num].ymin) {
+            update_bb_edge(net_id,
+                           bb_edge_new,
+                           bb_coord_new,
+                           bb_pin_sink_count_new,
+                           curr_bb_edge[layer_num].ymin,
+                           curr_bb_coord[layer_num].ymin,
+                           bb_edge_new[layer_num].ymin,
+                           bb_coord_new[layer_num].ymin);
+            if (bb_updated_before[net_id] == GOT_FROM_SCRATCH) {
+                return;
+            }
+        }
+
+        if (y_new > curr_bb_coord[layer_num].ymax) {
+            bb_edge_new[layer_num].ymax = 1;
+            bb_coord_new[layer_num].ymax = y_new;
+        } else if (y_new == curr_bb_coord[layer_num].ymax) {
+            bb_edge_new[layer_num].ymax = curr_bb_edge[layer_num].ymax + 1;
+            bb_coord_new[layer_num].ymax = curr_bb_coord[layer_num].ymax;
+        }
+    }
+}
+
+static inline void update_bb_layer_changed(ClusterNetId net_id,
```

Comment:
It seems like update_bb_layer_changed and update_bb_same_layer might be mergable if you wrote the code as "remove from old location" and "insert in new location". But maybe that's less efficient; just something to think about ...